RD-3879 Change Deployment Button to Environment/Service Button

leon-good-life opened the pull request:
<!--- Provide JIRA issue ID and a general summary of your changes in the Title above -->
<!--- e.g. "RD-1234 Improve styling in SuperComponent" -->
<!--- Note that the title is used as a commit message after merging the PR -->

## Description
Jira ticket: https://cloudifysource.atlassian.net/browse/RD-3879
<!--- Describe your changes to help reviewers understand the details. -->

## Screenshots / Videos
Service Page:
![Screen Shot 2022-04-19 at 18 45 20](https://user-images.githubusercontent.com/17104212/164043452-b8ddbb5b-3e97-46e1-b2f3-da15dc1d938a.png)
Environments Page:
![Screen Shot 2022-04-19 at 18 45 29](https://user-images.githubusercontent.com/17104212/164043464-a842cd37-00ef-4ac3-955a-7c0e943b89b5.png)

Adding widget to a page using Edit Mode:
![Screen Shot 2022-04-20 at 12 29 46](https://user-images.githubusercontent.com/17104212/164197751-6ac99dd4-c5d4-4fd9-9594-6b8c2a4528f0.png)

Edit Mode:
![Screen Shot 2022-04-20 at 13 11 43](https://user-images.githubusercontent.com/17104212/164205656-5fe9c4aa-d244-498d-a13d-a6a0d39b09de.png)

<!--- If applicable, add screenshot(s) or video(s) describing the changes you made. -->
<!--- Consider adding comparison screenshot(s) - before and after the change. -->
<!--- If not applicable, just write “N/A” in this section. -->

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] My code follows the [UI Code Style](https://cloudifysource.atlassian.net/l/c/x8fq902p).
- [x] I verified that all tests and checks have passed.
- [x] I verified if that change requires any update in the [documentation](https://cloudifysource.atlassian.net/l/c/4XKtPocy).
- [x] I added proper labels to this PR.

## Tests

20 April 2022, I started system tests after addressing review comments:
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/2567/pipeline

System tests after the implementing the rest of the functionality: 
https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/2561/
All system tests passed

System test after implementing correctly:
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/2560/pipeline
Result: all tests passed.

System tests: https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/2556/pipeline
<!--- If applicable, describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If applicable, provide a link to system tests, e.g: -->
<!--- [#1039](https://jenkins.cloudify.co/.../1039) (2021-08-06 07:05:50) --> 
<!--- If not applicable, just write “N/A” in this section. -->

## Documentation
https://github.com/cloudify-cosmo/docs.getcloudify.org/pull/2174
<!--- If applicable, describe how you documented or plan to document your changes. -->
<!--- Check UI documentation guidelines - https://cloudifysource.atlassian.net/l/c/EYWJ0XfB - for details. -->
<!--- Provide links to JIRA issues, other PRs or related documentation pages. -->
<!--- If not applicable, just write “N/A” in this section. -->

kubama commented:
Is this PR complete yet? The story description says: "Create deployment" button label on the Environments page should be changed to "Create Environments" and on the Services page should be changed to “Create Services”.
I can see this PR contains only the first part.

leon-good-life commented:
@kubama I fixed the implementation according to the requirements (Environments Page and Service Page). I achieved that by adding a `buttonContent` `configuration` to the `widget`.

leon-good-life commented:
System test after implementing correctly:
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/2560/pipeline

leon-good-life commented:
All system tests passed

leon-good-life commented:
I started system tests again after implementing the rest of the functionality:
https://jenkins.cloudify.co/view/UI%20Health%20View/job/Stage-UI-System-Test/

leon-good-life commented:
@kubama  I added:

>  icon, color, label and basic (similar to what we have in Service Button widget)

and I updated the documentation.

kubama commented on widgets/deploymentButton/README.md:
```diff
@@ -6,4 +6,8 @@ Allows creating a deployment of a chosen blueprint. After choosing a name for th
+* `label` - The label displayed in the button. Default: 'Create deployment'
+* `Color` - The color of the button. Available colors list can be found
+  at: [Theming - Semantic UI React](https://react.semantic-ui.com/layouts/theming). Default: 'green'
+* `icon` -  Name of the icon displayed in the button. Available icons list can be found
```
Can be found where?

leon-good-life replied:
I fixed the description: [78a02df](https://github.com/cloudify-cosmo/cloudify-stage/pull/1931/commits/78a02dfe5c77e9971e7183d64310e94382c7e445). I fixed also in docs repo.

kubama commented on widgets/deploymentButton/src/DeploymentButton.tsx:
```diff
@@ -1,46 +1,34 @@
 // @ts-nocheck File not migrated fully to TS
```
It should be easy to complete the migration, can we give it a try?

leon-good-life replied:
I migrated to TypeScript: c343057

kubama commented on widgets/deploymentButton/src/widget.tsx:
```diff
@@ -1,6 +1,9 @@
 // @ts-nocheck File not migrated fully to TS
```
It should be easy to complete the migration, can we give it a try?

leon-good-life replied:
I migrated to TypeScript: c343057

kubama commented on widgets/deploymentButton/README.md:
```diff
@@ -6,4 +6,9 @@ Allows creating a deployment of a chosen blueprint. After choosing a name for th
+* `Color` - The color of the button. Available colors list can be found
+  at: [Theming - Semantic UI React](https://react.semantic-ui.com/layouts/theming). Default: 'green'
+* `Icon` - Name of the icon displayed in the button. Available icons list can be found
+  at: [Icon - Semantic UI React](https://react.semantic-ui.com/elements/icon). Default: 'external'
```
Looks like the default is 'rocket', not 'external'

leon-good-life replied:
I fixed [5c409dc](https://github.com/cloudify-cosmo/cloudify-stage/pull/1931/commits/5c409dcc181f12776971202f91599aa86627e385), I fixed it also in docs repo.

kubama commented on app/translations/en.json:
```diff
@@ -1356,6 +1356,29 @@
             }
         },
+        "deploymentButton": {
+            "name": "Deployment button",
+            "description": "Displays button allowing to create a new service",
```
Looks like these translation keys are not used anywhere

leon-good-life replied:
I updated the translation and I'm now using it: 
7b0cadf

kubama commented on widgets/deploymentButton/src/widget.tsx:
```diff
@@ -1,6 +1,9 @@
+const t = Stage.Utils.getT(`widgets.${widgetId}`);
+
 Stage.defineWidget({
     id: 'deploymentButton',
```
Can we use `widgetId` here?

leon-good-life replied:
I changed to `widgetId`: [c4f8b52](https://github.com/cloudify-cosmo/cloudify-stage/pull/1931/commits/c4f8b521716c38ec49abf5450330214af485d8f7)

kubama commented on widgets/deploymentButton/src/widget.tsx:
```diff
@@ -11,10 +14,43 @@ Stage.defineWidget({
+            type: Stage.Basic.GenericField.STRING_TYPE
+        }
+    ],
     permission: Stage.GenericConfig.WIDGET_PERMISSION('deploymentButton'),
```
Can we use `widgetId` here?

leon-good-life replied:
I changed to `widgetId`: [a08fcbf](https://github.com/cloudify-cosmo/cloudify-stage/pull/1931/commits/a08fcbf284319d2a0fb21ee3df74d8c7569ccb97)

kubama commented on widgets/deploymentButton/src/widget.tsx:
```diff
@@ -1,20 +1,64 @@
+    initialConfiguration: [
+        {
+            id: 'basic',
+            name: t('configuration.basic.name'),
+            description: t('configuration.basic.description'),
+            default: false,
+            type: Stage.Basic.GenericField.BOOLEAN_TYPE
+        },
+        {
+            id: 'color',
+            name: t('configuration.color.name'),
+            description: t('configuration.color.description'),
+            default: 'green',
+            component: Stage.Common.Components.SemanticColorDropdown,
+            type: Stage.Basic.GenericField.CUSTOM_TYPE
+        },
+        {
+            id: 'icon',
+            name: t('configuration.icon.name'),
+            description: t('configuration.icon.description'),
+            default: 'rocket',
+            component: Stage.Shared.SemanticIconDropdown,
+            type: Stage.Basic.GenericField.CUSTOM_TYPE
+        },
+        {
+            id: 'label',
+            name: t('configuration.label.name'),
+            description: t('configuration.label.description'),
+            default: t('configuration.label.default'),
```
Maybe worth to add `tConfiguration` shorthand function to avoid repeating `configuration.` part of the translation key?

leon-good-life replied:
I added `tConfiguration`: [a08a095](https://github.com/cloudify-cosmo/cloudify-stage/pull/1931/commits/a08a095b79185054abac513dd5abe2ae7cd3ca2f)

leon-good-life replied:
And in [5efd5d0](https://github.com/cloudify-cosmo/cloudify-stage/pull/1931/commits/5efd5d0bdb6068db3a96a3e85bc82bd20ca6ab43)

kubama commented on widgets/deploymentButton/src/DeploymentButton.tsx:
```diff
@@ -1,46 +1,33 @@
+                className="widgetButton"
+                onClick={showModal}
+            />
+            <DeployBlueprintModal open={isModalOpen} onHide={hideModal} toolbox={toolbox} blueprintId={''} />
```
Can we make `blueprintId` prop optional instead of passing empty string?

leon-good-life replied:
I made `blueprintId` optional.

leon-good-life replied:
b7c481b
867be87

leon-good-life commented:
I started system tests after addressing review comments:
https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/2567/pipeline

kubama commented on widgets/common/src/deployModal/DeployBlueprintModal.tsx:
```diff
@@ -10,12 +10,14 @@ import { getErrorObject } from '../inputs/utils/errors';
->;
+    'open' | 'onHide' | 'toolbox'
+> {
+    blueprintId?: string;
```
I think it would be cleaner to make it optional in `GenericDeployModal` which already defines default value for this prop.
`DeployBlueprintModalProps` just passes the prop down, so it shouldn't define if it's required or not - it should accept whatever the target component accepts.

leon-good-life replied:
[3a96e98](https://github.com/cloudify-cosmo/cloudify-stage/pull/1931/commits/3a96e98f1b5aebd41aef245706ec1849f79feda5)

leon-good-life commented:
All system tests passed

kubama reviewed (approved):
Looks good, thanks for taking extra effort for TS migrations.
